### DISCUSSION TITLE
根据用户需求，在后台动态创建控件，这是不是违背mvvm原则， 是否还有更好的处理方式

### DISCUSSION THREAD

**Discussion by Vince0911** [BYJRK/DotNet-Discussions#207] [MVVM]
hi，
业务场景中，需要根据用户需求或配置，在后台动态创建控件，例如Image显示图片控件，用户可以根据需求，自己增加或减少Image的数量，能想到的方式就是后台创建一个ObservableCollection<Image>的属性，前台ItemsControl绑定这个属性，但是这样感觉控件是由后台创建的，是不是违背了mvvm原则？

**Comment by BYJRK** [comment]
这并不违背。否则按照这个逻辑，`ObservableCollection` 根本就不应该被用在 MVVM 当中，毕竟前台空间都是根据 VM 中的集合来动态生成的。真正违背的方式是，你把 `ItemsControl` 的引用传给 VM，然后在 VM 里面根据集合中的元素创建 Item 控件

**Reply by xingzuhui** [reply]
其实,我看到导航服务和dialog服务,被写在了vm里,不知道是不是违背原则了,prism的区域也写在vm里.
vm只能操作数据的话,很多前端逻辑xaml实现不了,必须行为或者xaml.cs里写.
